### DISCUSSION TITLE
GET support for user verification?

### DISCUSSION THREAD

**Discussion by jdukewich** [fastapi-users/fastapi-users#714] [Ideas]
A common method for user verification includes sending an email with the verification token, however I find it a lot more convenient if then the `/verify` endpoint handled GET requests instead of POST requests. For example, I think it would be easier to format a text email with the JWT encoded in the URL as a parameter instead of attempting to write an HTML email template with a form to allow the post request. Would this make sense? If so, could there be an additional parameter for the verify router to say whether you want to verify route as a GET or POST?

**Comment by bittermandel** [comment]
This does generally not work as email providers often send a GET request to all URLs embedded in emails. Perhaps a URL that sends the user to a page where they can submit a form could work, but then it's just working around the problem of sending HTML emails.

**Reply by jdukewich** [reply]
Okay, thanks for your reply. That makes sense. I'll just bite the bullet and work with HTML templates then.

**Reply by jdukewich** [reply]
Sorry to reopen this, but I'm confused why this is the case. I have noticed plenty of times (e.g. Verizon verification email) where the button to verify an email is actually just an <a href=""> so why can all these other sites use GET requests but we can't in this instance?

**Comment by bert2002** [comment]
I think that would be a great idea and is quite common practice, then the user does not need to copy and paste the verification token and the on-boarding process is much smoother, e.g. the user gets an email and just needs to click (as an example):

```
http://127.0.0.1:8000/auth/verify?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiOGVkYTJlZDQtZjY5My00MjI5LWEwMmYtODM3ODM3ODkxZjc5IiwiZW1haWwiOiJzLndpcnRoKzIwMzgxQGl0YmVydC5kDSIsImF1ZCI6ImZhc3RhcGktdXNlcnM6dmVyaWZ5IiwiZXhwIjoxNjMzNzEwNzQzfQ.g9j-BsobYF-nROUSGHJtWR4dZ6NuVL1-KSdGrI97L38
```

**Reply by frankie567** [reply]
In the context of a REST API, which is the use-case FastAPI Users target, it's better to implement it using a `POST` request. If you need to it on the frontend side, you just have to implement a simple JavaScript function making an AJAX request to the endpoint with the JSON payload.

**Reply by jdukewich** [reply]
Implementing JavaScript to make the AJAX request will not work very well if we wanted to do an email verification. I'd like to be able to send an email where the user can click a link to verify. This is more useful if this is implemented as a GET. This could be implemented as a POST if we embed a form within the email, but then that leads to CORS issues and email providers often have security measures preventing you from making that cross site request.

**Reply by frankie567** [reply]
You can redirect to a GET page on your frontend application. This page can very well have some JS code that triggers an AJAX POST request to the API with the required parameters, directly when the page is loaded.

**Reply by KentonParton** [reply]
@frankie567 would you be against me adding a GET equivalent?

I have added the following to the "verify" router [here](https://github.com/fastapi-users/fastapi-users/blob/c759bb691569306336e099a976aa51e352322143/fastapi_users/router/verify.py#L86) and works like a charm. This will have no breaking changes.

```python
    @router.get("/verify", response_model=user_model)
    async def verify(
        request: Request,
        token: str,
        user_manager: BaseUserManager[models.UC, models.UD] = Depends(get_user_manager),
    ):
        try:
            return await user_manager.verify(token, request)
        except (InvalidVerifyToken, UserNotExists):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=ErrorCode.VERIFY_USER_BAD_TOKEN,
            )
        except UserAlreadyVerified:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=ErrorCode.VERIFY_USER_ALREADY_VERIFIED,
            )
```

This would allow users to get verified by clicking on links sent to them via email.

**Reply by frankie567** [reply]
Yes, I'm against it. Indeed, it would totally work by allowing the `GET` method, however, it would remain a REST API with a JSON response. For an end-user clicking on a e-mail link, that would be weird and incomprehensible.

Once again, my goal with FastAPI Users is to provide a REST API, not a static HTML page renderer. My recommendation to implement the verify action is to implement a page on your frontend application (accessible by a GET request, so a user can click on the link) and perform the verification with an AJAX call. This way, we can easily:

* Handle the errors and show proper error messages to the user
* Show a success message
* Redirect the user to another relevant page to get them started
* ...

If one really need to handle it on the backend side, I recommend to implement a custom route to fit their needs. With the UserManager, you only need to call the `.verify` method. Then you can for example render an HTML template or whatever.

**Reply by KentonParton** [reply]
Makes total sense. Thanks for explaining the short-comings of handling it on the server side.
